src/erasure.rs

```diff
@@ -0,0 +1,441 @@
+            let mut coding_ptrs: Vec<&[u8]> = Vec::new();
+            let mut erasures: Vec<i32> = Vec::new();
+            for i in consumed..coding_end {
+                let j = i % window.len();
+                let mut b = &mut window[j];
+                if b.is_some() {
+                    blobs.push(b.clone().unwrap());
+                    continue;
+                }
+                let n = re.allocate();
+                *b = Some(n.clone());
+                //mark the missing memory
+                blobs.push(n);
+                erasures.push((i - consumed) as i32);
+            }
+            erasures.push(-1);
+            trace!("erasures: {:?}", erasures);
+            //lock everything
+            for b in blobs.iter() {
+                locks.push(b.write().unwrap());
+            }
+            for (i, l) in locks.iter_mut().enumerate() {
+                if i >= NUM_DATA {
+                    trace!("pushing coding: {}", i);
+                    coding_ptrs.push(&l.data);
+                } else {
+                    trace!("pushing data: {}", i);
+                    data_ptrs.push(&mut l.data);
+                }
+            }
+            trace!(
+                "coding_ptrs.len: {} data_ptrs.len {}",
+                coding_ptrs.len(),
+                data_ptrs.len()
+            );
+            decode_blocks(data_ptrs.as_mut_slice(), &coding_ptrs, &erasures)?;
+        } else {
+            return Err(ErasureError::NotEnoughBlocksToDecode);
+        }
+    }
+    Ok(())
+}
+
+#[cfg(test)]
+mod test {
+    use erasure;
+    use packet::{BlobRecycler, SharedBlob, PACKET_DATA_SIZE};
+    extern crate env_logger;
+
+    #[test]
+    pub fn test_coding() {
+        let zero_vec = vec![0; 16];
+        let mut vs: Vec<Vec<u8>> = Vec::new();
+        let v_orig: Vec<u8> = vec![0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15];
+        for i in 0..4 {
+            let mut v = v_orig.clone();
+            for mut x in v.iter_mut() {
+                *x += i;
+            }
+            vs.push(v);
+        }
+
+        let m = 2;
+        let mut coding_blocks = Vec::new();
+        for _i in 0..m {
+            let new_block = vec![0u8; v_orig.len()];
+            coding_blocks.push(new_block);
+        }
+
+        {
+            let mut coding_blocks_slices = Vec::new();
+            for mut coding_block in coding_blocks.iter_mut() {
+                coding_blocks_slices.push(coding_block.as_mut_slice());
+            }
+            let mut v_slices = Vec::new();
+            for mut v in vs.iter() {
+                v_slices.push(v.as_slice());
+            }
+
+            assert!(
+                erasure::generate_coding_blocks(
+                    coding_blocks_slices.as_mut_slice(),
+                    v_slices.as_slice()
+                ).is_ok()
+            );
+        }
+        trace!("coding blocks:");
+        for b in coding_blocks.iter() {
```

Comment:
In Rust, it's more idiomatic to write `&coding_blocks` instead of `coding_blocks.iter()` in for loops.